### DISCUSSION TITLE
ffprobe HLS "Could not find codec parameters for stream"

### DISCUSSION THREAD

**Discussion by JonathanThorpe** [savonet/liquidsoap#3809] [User Support]
I'm attempting to set up a HLS stream that comprises three bitrates using Liquidsoap 2.2.4 (from the official Docker container): 128Kbps AAC, 64Kbps HE-AAC and 32Kbps HE-AAC.

```
# HLS config
config_aac_lofi = %ffmpeg(format="adts",
                  %audio(
                    codec="libfdk_aac",
                    channels=2,
                    samplerate=48000,
                    b="32k",
                    afterburner=1,
                    profile="aac_he"
                  )).{
                      id3_version = 3,
                      codecs="mp4a.40.5"
                    }

config_aac_midfi = %ffmpeg(format="adts",
                  %audio(
                    codec="libfdk_aac",
                    channels=2,
                    samplerate=48000,
                    b="64k",
                    afterburner=1,
                    profile="aac_he"
                  )).{
                      id3_version = 3,
                      codecs="mp4a.40.5"
                    }

config_aac_hifi = %ffmpeg(format="adts",
                  %audio(
                    codec="aac",
                    channels=2,
                    ar=48000,
                    b="128k"
                  )).{
                      id3_version = 3,
                      # required to pass the Apple HLS Validation Report
                      bandwidth=150000
                    }

# Stream to HLS

output.file.hls(playlist="#{station_hls}.m3u8",
                hls_segments_directory,
                temp_dir=temp_dir,
                persist_at="#{persist_dir}/persist.config",
                segment_duration=4.,
                segments=7,
                segments_overhead=10,
                segment_name=segment_name,
                extra_tags = ["#EXT-X-INDEPENDENT-SEGMENTS"],
                [("hifi", config_aac_hifi),
                   ("midfi", config_aac_midfi),
                   ("lofi", config_aac_lofi)],
                hls_source)
```

It works for the most part, however while troubleshooting an app that consumes this stream (plays for 30s and then stops before starting again - not network related), I noticed this output from ffprobe:

    ffprobe -i https://streaming-hls.2me.com.au/hls/2me/live/all/2me.m3u8
    ffprobe version 6.1.1 Copyright (c) 2007-2023 the FFmpeg developers
    ...
    [hls @ 0x1308042d0] Skip ('#EXT-X-VERSION:3')
    [hls @ 0x1308042d0] Skip ('#EXT-X-INDEPENDENT-SEGMENTS')
    [hls @ 0x1308042d0] Opening 'https://streaming-hls.2me.com.au/hls/2me/live/all/aac_hifi.m3u8' for reading
    [hls @ 0x1308042d0] Skip ('#EXT-X-VERSION:3')
    [hls @ 0x1308042d0] Skip ('#EXT-X-DISCONTINUITY-SEQUENCE:4')
    [https @ 0x13102f400] Opening 'https://streaming-hls.2me.com.au/hls/2me/live/all/aac_midfi.m3u8' for reading
    [hls @ 0x1308042d0] Skip ('#EXT-X-VERSION:3')
    [hls @ 0x1308042d0] Skip ('#EXT-X-DISCONTINUITY-SEQUENCE:0')
    [https @ 0x13102f400] Opening 'https://streaming-hls.2me.com.au/hls/2me/live/all/aac_lofi.m3u8' for reading
    [hls @ 0x1308042d0] Skip ('#EXT-X-VERSION:3')
    [hls @ 0x1308042d0] Skip ('#EXT-X-DISCONTINUITY-SEQUENCE:0')
    [hls @ 0x1308042d0] Opening 'https://streaming-hls.2me.com.au/hls/2me/live/all/aac_aac_hifi_4_30280_1711057144.aac' for reading
    [hls @ 0x1308042d0] Opening 'https://streaming-hls.2me.com.au/hls/2me/live/all/aac_aac_hifi_4_30281_1711057154.aac' for reading
    [hls @ 0x1308042d0] Opening 'https://streaming-hls.2me.com.au/hls/2me/live/all/aac_aac_midfi_4_30280_1711057144.aac' for reading
    [hls @ 0x1308042d0] Opening 'https://streaming-hls.2me.com.au/hls/2me/live/all/aac_aac_midfi_4_30281_1711057154.aac' for reading
    [hls @ 0x1308042d0] Opening 'https://streaming-hls.2me.com.au/hls/2me/live/all/aac_aac_lofi_4_30280_1711057144.aac' for reading
    [hls @ 0x1308042d0] Opening 'https://streaming-hls.2me.com.au/hls/2me/live/all/aac_aac_lofi_4_30281_1711057154.aac' for reading
    [hls @ 0x1308042d0] Could not find codec parameters for stream 0 (Audio: aac, 0 channels, fltp): unspecified sample rate
    Consider increasing the value for the 'analyzeduration' (0) and 'probesize' (5000000) options
    [hls @ 0x1308042d0] Could not find codec parameters for stream 1 (Audio: aac, 0 channels, fltp): unspecified sample rate
    Consider increasing the value for the 'analyzeduration' (0) and 'probesize' (5000000) options
    Input #0, hls, from 'https://streaming-hls.2me.com.au/hls/2me/live/all/2me.m3u8':
      Duration: N/A, bitrate: N/A
      Program 0 
        Metadata:
          variant_bitrate : 150000
      Stream #0:0: Audio: aac, 0 channels, fltp
        Metadata:
          variant_bitrate : 150000
          title           : News
          artist          : VDL
          id3v2_priv.?com.apple.streaming.transportStreamTimestamp: \x00\x00\x00\x001\xd0\x7f\xf3
      Program 1 
        Metadata:
          variant_bitrate : 70400
      Stream #0:1: Audio: aac, 0 channels, fltp
        Metadata:
          variant_bitrate : 70400
          title           : News
          artist          : VDL
          id3v2_priv.?com.apple.streaming.transportStreamTimestamp: \x00\x00\x00\x001\xd0\x7f\xf3
      Program 2 
        Metadata:
          variant_bitrate : 35200
      Stream #0:2: Audio: aac (HE-AAC), 48000 Hz, stereo, fltp
        Metadata:
          variant_bitrate : 35200
          title           : News
          artist          : VDL
          id3v2_priv.?com.apple.streaming.transportStreamTimestamp: \x00\x00\x00\x001\xd0\x7f\xf3

Note that Program 1 and Program 2 don't show the right codec and says `Stream #0:1: Audio: aac, 0 channels, fltp`. This is despite Program 2 using the same codec as Program 1 (just a lower bitrate).

There is also this in the output:

    [hls @ 0x1308042d0] Could not find codec parameters for stream 0 (Audio: aac, 0 channels, fltp): unspecified sample rate
    Consider increasing the value for the 'analyzeduration' (0) and 'probesize' (5000000) options
    [hls @ 0x1308042d0] Could not find codec parameters for stream 1 (Audio: aac, 0 channels, fltp): unspecified sample rate
    Consider increasing the value for the 'analyzeduration' (0) and 'probesize' (5000000) options

And finally the duration and bitrate are N/A which doesn't seem right.

I've done as suggested by increasing analyzeduration and probesize to 100M which did not solve the issue.

If I run ffprobe on the individual streams contained within the master playlist, they report back properties ok, but still the overall bitrate is "N/A" (not sure if this is important):

    ffprobe -i https://streaming-hls.2me.com.au/hls/2me/live/hls/aac_lofi.m3u8 -hide_banner
    [hls @ 0x10f604370] Skip ('#EXT-X-VERSION:3')
    [hls @ 0x10f604370] Skip ('#EXT-X-DISCONTINUITY-SEQUENCE:0')
    [hls @ 0x10f604370] Opening 'https://streaming-hls.2me.com.au/hls/2me/live/hls/aac_aac_lofi_4_30354_1711057869.aac' for reading
    [hls @ 0x10f604370] Opening 'https://streaming-hls.2me.com.au/hls/2me/live/hls/aac_aac_lofi_4_30355_1711057879.aac' for reading
    Input #0, hls, from 'https://streaming-hls.2me.com.au/hls/2me/live/hls/aac_lofi.m3u8':
      Duration: N/A, bitrate: N/A
      Program 0 
        Metadata:
          variant_bitrate : 0
      Stream #0:0: Audio: aac (HE-AAC), 48000 Hz, stereo, fltp
        Metadata:
          variant_bitrate : 0
          title           : 1638AM
          artist          : Radio 2ME
          id3v2_priv.?com.apple.streaming.transportStreamTimestamp: \x00\x00\x00\x005\xb4>\xe3

To sanity check my ffprobe, I can successfully run this against another stream:

    ffprobe -i https://mediaserviceslive.akamaized.net/hls/live/2038308/triplejnsw/index.m3u8
    <snip>
    Input #0, hls, from 'https://mediaserviceslive.akamaized.net/hls/live/2038308/triplejnsw/index.m3u8':
      Duration: N/A, start: 50021.764989, bitrate: 510 kb/s
      Program 0 
        Metadata:
          variant_bitrate : 211200
      Stream #0:0: Audio: aac (LC), 44100 Hz, stereo, fltp, 190 kb/s
        Metadata:
          variant_bitrate : 211200
          id3v2_priv.com.apple.streaming.transportStreamTimestamp: \x00\x00\x00\x01\x0cZ(Q
          id3v2_priv.com.elementaltechnologies.timestamp.utc: 2024-03-21T10:02:11Z
      Program 1 
        Metadata:
          variant_bitrate : 70400
      Stream #0:1: Audio: aac (HE-AAC), 44100 Hz, stereo, fltp, 63 kb/s
        Metadata:
          variant_bitrate : 70400
          id3v2_priv.com.apple.streaming.transportStreamTimestamp: \x00\x00\x00\x01\x0cVp\xc1
          id3v2_priv.com.elementaltechnologies.timestamp.utc: 2024-03-21T10:02:09Z
      Program 2 
        Metadata:
          variant_bitrate : 211200
      Stream #0:2: Audio: aac (LC), 44100 Hz, stereo, fltp, 192 kb/s
        Metadata:
          variant_bitrate : 211200
          id3v2_priv.com.apple.streaming.transportStreamTimestamp: \x00\x00\x00\x01+\x0f\x82p
          id3v2_priv.com.elementaltechnologies.timestamp.utc: 2024-03-21T11:37:36Z
      Program 3 
        Metadata:
          variant_bitrate : 70400
      Stream #0:3: Audio: aac (HE-AAC), 44100 Hz, stereo, fltp, 64 kb/s
        Metadata:
          variant_bitrate : 70400
          id3v2_priv.com.apple.streaming.transportStreamTimestamp: \x00\x00\x00\x01+\x10A\xbc
          id3v2_priv.com.elementaltechnologies.timestamp.utc: 2024-03-21T11:37:36Z

A few things to note:

1. The above stream and the one I'm attempting to fix both use ADTS muxing
2. I've ruled out web servers / proxies as the issue by downloading all of the playlists and segments and run ffprobe locally - same issue.
3. I took the master playlist from the working stream and updated the parameters to reflect my stream in case one of the parameters I don't have (i.e. "AVERAGE-BANDWIDTH") mattered.
4. I've tried both %fdkaac and %ffmpeg encoders.
5. I've also tried setting all streams to AAC LC.

Does anyone have any suggestions as to where I can go from here?

**Comment by JonathanThorpe** [comment]
Doing some further debugging on this, one thing that might be contributing to this could be the mandatory "id3v2_priv.com.apple.streaming.transportStreamTimestamp". Is there any reason this appears as `id3v2_priv.?com.apple.streaming.transportStreamTimestamp` (note the erroneous `?`)?

**Comment by toots** [comment]
I can confirm that the bitrate does show up on my `ffprobe` output with the fix for https://github.com/savonet/liquidsoap/issues/3817

**Reply by JonathanThorpe** [reply]
Thanks @toots  - I really appreciate your help!

**Reply by JonathanThorpe** [reply]
Just confirming: Using the rolling release, this fixes the ffprobe output and also the issue with multi-variant HLS streams stopping. Thanks again!

**Reply by toots** [reply]
Great! One weird side effect is that this broke in-stream metadata when listening with our ffmpeg backend. I have submitted a patch for it.
